### DISCUSSION TITLE
webui-go still uses civetweb somewhere?

### DISCUSSION THREAD

**Discussion by laoshaw** [webui-dev/webui#489] [Q&A] [answered]
I can 'go build' examples from webui-go, e.g. serve_a_folder.

ldd serve_a_folder:
	linux-vdso.so.1 (0x00007ffe9f1fb000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f16bb200000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f16bb5ec000)

I wonder where 'civetweb' is, does webui-go uses it as a websocket server, or it creates its own websocket and just uses webui.js from the webui project?

**Comment by AlbertShown** [comment]
> I wonder where 'civetweb' is

`civetweb` is embedded inside `webui-2.so` / `webui-2-static.a`

> does webui-go uses it as a websocket server

Yes, webui use civetweb for web server and websocket server, and webui core is dealing with running web browsers, WebView, bind UI with backend etc...

**Reply by laoshaw** [reply]
Thanks for the reply. as far as webui's go-binding goes, can I use all Go APIs freely? or are they limited by webui's bindings?  if all Go api can be used then it has a much better cross platform experience than c or c++ thanks to Go's cross platform design(assuming no cgo needed that is)

**Reply by AlbertShown** [reply, marked as answer]
You can use all complete Go APIs.
This is not because webui binding, but because webui simply sends an event from UI to your backend Go function, then can use all complete Go APIs as you like, then sends back a response to UI.

**Comment by laoshaw** [comment]
This brings up an interesting question, what about using a golang-websocket to replace civetweb, since Go is more cross-platform friendly than c? in that case webui could become a single go-binary, similar to Wails but using browser and webview(Wails can only do webview), with Go the deployment will be much easier, but Go might be harder to provide binding to other languages, C is the best there for multi-language binding.

**Reply by AlbertShown** [reply]
> but Go might be harder to provide binding to other languages

This is the main reason why webui is written in C I guess :)
